sys/mathoid.js

```diff
@@ -0,0 +1,140 @@
+'use strict';
+
+
+var P = require('bluebird');
+var HyperSwitch = require('hyperswitch');
+var URI = HyperSwitch.URI;
+var HTTPError = HyperSwitch.HTTPError;
+
+
+function MathoidService(options) {
+
+    this.options = options;
+
+}
+
+
+MathoidService.prototype.checkInput = function(hyper, req) {
+
+    var self = this;
+    var rp = req.params;
+    var hash;
+    var origHash;
+    var checkRes;
+
+    // start by calculating the hash
+    return hyper.post({
+        uri: new URI([rp.domain, 'sys', 'post_data', 'mathoid.input', 'hash']),
+        body: { q: req.body.q, type: rp.type }
+    }).then(function(res) {
+        hash = origHash = res.body;
+        // short-circuit if it's a no-cache request
+        if (req.headers && /no-cache/.test(req.headers['cache-control'])) {
```

Comment:
We still need the original hash later at L80 to know whether we need to store an indirection record or not.